### ISSUE TITLE
Tracing in background

### ISSUE THREAD

**Issue by UsamaAbdulAziz** [c19x/C19X-iOS#2]
Amazing work.
There is this issue, I am unable to trace contacts when both iPhones with the app are on lockscreen (or just app is in background) and when I bring both phones next to each other.
There is no increment in contacts in such case. Only when I bring app to foreground, it resumes.

**Comment by c19x** [comment]
Thanks for your kind words. I broke the beacon code recently while trying to optimise and reduce the number of connects, so I have also seen this issue which was intermittent in my case. Sorry for pushing a buggy update. I shall provide an update today. More generally, I am still unhappy with CoreBluetooth reliability when both phones are in suspended state. In particular, if two phones that have never seen each other and haven't detected anything for a long time, they will both enter suspended or even killed state. If you bring them together, they SHOULD launch on detection by CoreBluetooth as described in Apple documentation, but they only do that sometimes. I have already added BGAppRefresh and BGProcessing tasks in AppDelegate to compensate for this scenario, but even these don't really behave that reliably (BGAppRefresh is not called that regularly or even at all). As such, my only option is to incorporate device movement detection as trigger, which is working very well on my device right now, so I shall optimise and push to master this week.

**Comment by UsamaAbdulAziz** [comment]
That’s great. 
Yes I totally agree. I have been facing the same issues but you have handled almost all of those cases. 
I have learnt a lot from your work and I am now looking forward to this new commit. Keep up your awesome work! 👍

**Comment by c19x** [comment]
Hello! The device movement detection update is now on GitHub, along with a few other tweaks. Yes, CoreBluetooth and BGAppRefresh don't behave as advertised, otherwise this would have been easy. Location manager has now been added which triggers app on Visits and Significant Location changes. This means when both phones that have never seen each other are in suspended mode, as long as one of them has moved significantly or started moving again after a period of being stationary, it SHOULD wake up and start BLE scan which in turn should wake the other phone up when in range🤞I am trying to avoid using fine grain location updates to preserve battery life, but that may be necessary if this isn't working as advertised by Apple 🤦🏻‍♂️ You will see these other options are already available in LocationManager, but set to false at the moment, feel free to experiment.

**Comment by UsamaAbdulAziz** [comment]
Perfect. That’s a good idea to use location for BLE scanning when app has not been used for some time. In theory, there won’t be any need for this if BLE worked as advertised. I will test this and also experiment a few things if any case fails.
Will test it and will report if there is any possibility for improvement. Your work is a masterpiece and great help for devs in these difficult times. 👍

**Comment by UsamaAbdulAziz** [comment]
Just so you know, I tested the latest commit by installing on 2 iPhones and moved one away for good 6 7 hours. 
I put on a user notification to know whenever the application is awake due to location. 
Also notification when contact establishes. 
As I returned back to the stationary iPhone I received a location notification and visits notification as well. That means that the application in moving iPhone was there in background.
Despite this, I did not get any contact. I waited for almost 10 minutes keeping both iPhones together. 
As soon as I opened the app, hence changing moving iPhone’s state to foreground, contact established immediately. 

I am trying to debug the issue on my end as well. Just wanted to let you know about this test.

**Comment by c19x** [comment]
Thanks for sharing the test results. I've had some minor success but also found that CoreLocation doesn't seem to help at the moment 😕. I have a few ideas to test, shall keep you posted. I've also uploaded my dev branch "LocationManager" which incorporates much improved CoreBluetooth state restoration code and app terminate / crash simulation. It proves a terminated app on phone A can definitely resume on detection of another phone B in background / suspended state, but not in terminated state yet.

Test procedure
Phone A will act as terminated App, Phone B will act as trigger.
Use Console to view log (switch on Action > Include Info / Debug messages, then filter by subsystem org.c19x), rather than Xcode. Start app on phone rather than Xcode, as app termination while connected to Xcode will enter debug mode.
1. Phone B, switch on Airplane mode, make sure Bluetooth connectivity has been disabled too.
2. Phone B, start App, move it to background, lock phone. App doesn't do anything as Bluetooth is OFF.
3. Phone A, make sure Bluetooth is ON. Start App, shouldn't detect anything as Phone B is not broadcasting.
4. Phone A, double tap on Recent Contacts box, dialog asks "Delete Contacts", press Cancel to trigger app crash in 5 seconds. The App should disappear on its own, this is equivalent to OS termination of app. Console log will show "simulatedCrash" as fault message.
5. Phone B, app should be in suspended state by now. Use control centre to switch off Airplane mode and ensure Bluetooth is also ON too. Console should show App on Phone B as broadcasting.
6. View console messages for Phone A, app should have resumed automatically, starting with Database load, Controller init, etc and then beacon detection 👍

This has worked every time, at last 😵 So some minor success. So the last challenge is detection between two terminated apps.

Weirdly, I have also observed one instance of beacon resume after going out of range and returning again. But haven't been able to replicate again this morning. I think it was just lucky.

[Beacon Test - 20200624-1900.pdf](https://github.com/c19x/C19X-iOS/files/4830369/Beacon.Test.-.20200624-1900.pdf)

**Comment by UsamaAbdulAziz** [comment]
Checked out “LocationManager” branch, and followed your defined test procedure. Your work here is marvellous. 
Only thing I am concerned about is when I turned bluetooth “ON” on the trigger phone, the app restarts on the second phone. Until this, everything’s perfect.
After this, there is an error logged in Console (error that is occurred when beacon[uuid] is null). Then there is an indefinite loop of same few lines of log and then error, where it tries to connect to peripheral in a loop. 
Don’t know if am doing anything wrong, because it stops app from establishing contact.

I faced this issue every time. 
Have a look:

![image](https://user-images.githubusercontent.com/22097793/85717185-e8f78600-b6f5-11ea-9c45-2c08420317ea.jpeg)

**Comment by c19x** [comment]
Hello! Thanks for testing the code. App restart is the correct behaviour because state restoration works as follows ...

1. When the app is terminated, CoreBluetooth takes over responsibility of handling any pending actions (e.g. connect, read, write). Search for `c19x` in console for process `bluetoothd` and you should see it is still working on the app's behalf.
2. When CoreBluetooth detects the action can now be completed (e.g. the other phone is in range), it will relaunch the app from fresh, albeit entering background state immediately, calling the `willRestoreState` function to give the app opportunity to recreate the last known state, and then calling the pending CoreBluetooth delegate method, which in this case is usually:
   
   - `peripheral(didWriteValueFor characteristic)` for a known iOS device
   - `centralManager(didDiscover)` for Android devices or NEW iOS device

3. The app then takes over responsibility for normal running, until it gets terminated again.

This is an incredibly fiddly and delicate process, with vital code fragments that are only mentioned in passing in Apple documentation (e.g. must maintain strong reference to restored peripherals, and take care not to overwrite peripheral advert 😬).

I've also seen the issue you mentioned on last night's code version, where a restored peripheral enters the unknown, connect, expired, disconnect loop. That has now been fixed and the dev version pushed to the LocationManager branch for trial 👍. I've also improved the logic to avoid queuing lots of pending `wakeTransmitter` calls, as just one is enough.

I think this is the status of the current beacon code:
F = Foreground
B = Background
S = Suspended
T = Terminated

- Android Central - Android Peripheral : Reliable
- Android Central - iOS Peripheral FBST : Reliable
- Android Peripheral - iOS Central FBST : Reliable
- iOS Central FBST - iOS Peripheral FB : Reliable
- iOS Central FB - iOS Peripheral FBST : Reliable

The focus now is getting to grips with the following scenario. 

- iOS Central ST - iOS Peripheral ST : ???

I am waiting for a couple of faraday bags to accelerate testing, as moving in and out of range is not sustainable. As an aside, if an Android phone is introduced, it solves the scenario, but that can not be a working assumption for practical deployments.

I am surprised location detection doesn't seem to do anything to help CoreBluetooth. It's almost as if the code is called but CB functions are mocked, weird.

**Comment by c19x** [comment]
Another little step forward. Latest commit on Master branch detects *NEW* iOS devices while in suspended / terminated state. The receiver state machine has also been refined to remove redundant readCode / readRSSI calls which are harmless but wasteful. LocationManager has been removed as that doesn't seem to help. Weirdly, even BGAppRefresh doesn't seem to help.

The only remaining issue now is connection resume for a pair of known iOS devices, i.e. the following scenario ...

1. iPhone A and iPhone B meet for the first time, connection established
2. The two phones keep each other awake
3. iPhone A moves out of range of iPhone B, triggers timeout error caught in `didDisconnectPeripheral` on iPhone B, which immediately issues a `connect`, and the timer will also issue a `scan` soon after that.
4. iPhone A moves back in range of iPhone B (now in suspended / terminated state), it should trigger iPhone B to wake up and handle the pending `connect` or `scan`, but it doesn't, hence the connection is broken.
5. Manually bringing app to foreground on iPhone A or B resolves the issue.

Apart from this situation, all other scenarios seem to be working well and stable. Faraday bags have turned up, that should help repeatable testing of this scenario to find root cause.

**Comment by UsamaAbdulAziz** [comment]
Hi.
I have been testing your code with a lot of scenarios, going out and coming back after some time. It works just as expected! 👍🤟
Playing with locationManager, but somehow it is designed to NOT assist CoreBluetooth operations. That’s a shame!
I have tested these scenarios and all of them work fine. 
I now need to test the NEW iOS in ST states that you have just pushed. 
Re-detection of known peripherals is essential in our case. I will be testing a few things that are on my mind to make it work in case of known devices in ST state. 
Will keep it posted. 👍👍

**Comment by c19x** [comment]
Thank you for all your efforts. It is much appreciated. The ability for iOS Central in ST state to detect NEW iOS peripheral is a little step forward, but you are right, it is essential to achieve iOS Central in ST state detecting KNOWN iOS peripheral in ST state. Good luck! If we crack this last problem, it will be the first complete solution that does not require iOS update. As such it will work on 81% of all phones today (rather than just 2% of iPhones on the latest OS version).

**Comment by c19x** [comment]
I might have found the problem and solution🤞Please pull latest commit on Master branch to try out.

It turns out the existing code works perfectly under FBST states for reconnect after out of range, except when one of the phones changes identity while the other is in S/T state, therefore the pending connect from Central is waiting on a Peripheral UUID that is no longer recognised. However, the pending write on the Peripheral is correctly received, and therefore it offers a mechanism for providing a scan "hint" for the new Peripheral UUID.

Here is an example of this situation on the previous code ...

1. Device gone out of range (UUID = 6F2367C5-C82C-0993-EF1D-3143BE24EC8D)
`2020-06-27 10:07:29.426865+0100 C19X[12140:5606801] [Receiver] didDisconnectPeripheral (peripheral=6F2367C5-C82C-0993-EF1D-3143BE24EC8D,error=Optional(Error Domain=CBErrorDomain Code=6 "The connection has timed out unexpectedly." UserInfo={NSLocalizedDescription=The connection has timed out unexpectedly.}))`
2. Pending connect for device (UUID = 6F2367C5-C82C-0993-EF1D-3143BE24EC8D)
`2020-06-27 10:07:29.427524+0100 C19X[12140:5606801] [Receiver] connect (source=didDisconnectPeripheral,peripheral=6F2367C5-C82C-0993-EF1D-3143BE24EC8D)
2020-06-27 10:07:29.432411+0100 C19X[12140:5606801] [Transmitter] Unsubscribe (central=6F2367C5-C82C-0993-EF1D-3143BE24EC8D)`
3. Pending scan for all devices and various scheduled tasks
`2020-06-27 10:07:37.428438+0100 C19X[12140:5606856] [Receiver] scan (source=scheduleScan|connect,statistics={count=38,mean=13.715828632053574,sd=5.860253064582239,min=8.051813006401062,max=32.66571402549744,period=521.2014880180359})
2020-06-27 10:07:37.529354+0100 C19X[12140:5606965] [Transmitter] Notify subscribers (source=didUnsubscribeFrom)
2020-06-27 10:07:37.541572+0100 C19X[12140:5606856] [Transmitter] Automatic beacon code update (lastUpdate=27/06/2020 10:05,elapsed=120.10971796512604)
2020-06-27 10:07:37.542195+0100 C19X[12140:5606856] [Transmitter] updateBeaconCode
2020-06-27 10:07:37.543580+0100 C19X[12140:5606856] [Transmitter] updateBeaconCode successful (code=-8954702307277676422,characteristic=0022D481-83FE-1F13-83BA-81B21157E07A)`
4. Device come back into range (UUID changed to 374D3A1A-4824-0186-B50F-09E87342DB5F)
`2020-06-27 10:09:31.016069+0100 C19X[12140:5606965] [Transmitter] Subscribe (central=374D3A1A-4824-0186-B50F-09E87342DB5F)
2020-06-27 10:09:39.010718+0100 C19X[12140:5606970] [Transmitter] Write (central=374D3A1A-4824-0186-B50F-09E87342DB5F)`
5. Transmitter loops indefinitely but beacon no longer detected as the pending connect is looking for a different UUID while pending scan only reports NEW devices but this is a known device.

Solution is to introduce code in transmitter, in 
`peripheralManager:didReceiveWrite`
to offer a hint to the receiver to look for a specific peripheral.
I might have to add this to other places in Transmitter, but it's a start.

Here is an example of latest code working ...

1. Device gone out of range (UUID = 131BB7C7-5001-3D03-6E8B-36911EC270D5)
`2020-06-27 11:35:27.121996+0100 C19X[12252:5638983] [Receiver] didDisconnectPeripheral (peripheral=131BB7C7-5001-3D03-6E8B-36911EC270D5,error=Optional(Error Domain=CBErrorDomain Code=6 "The connection has timed out unexpectedly." UserInfo={NSLocalizedDescription=The connection has timed out unexpectedly.}))`
2. Pending connect (UUID = 131BB7C7-5001-3D03-6E8B-36911EC270D5)
`2020-06-27 11:35:27.122396+0100 C19X[12252:5638983] [Receiver] connect (source=didDisconnectPeripheral,peripheral=131BB7C7-5001-3D03-6E8B-36911EC270D5)
2020-06-27 11:35:27.126946+0100 C19X[12252:5638983] [Transmitter] Unsubscribe (central=131BB7C7-5001-3D03-6E8B-36911EC270D5)`
3. Pending scan for all devices and scheduled tasks
`2020-06-27 11:35:35.194195+0100 C19X[12252:5638983] [Receiver] scan (source=scheduleScan|connect,statistics={count=6,mean=62.5800198316574,sd=79.90518350783763,min=8.845937013626099,max=217.46147894859314,period=375.48011898994446})
2020-06-27 11:35:35.197961+0100 C19X[12252:5639045] [Transmitter] Notify subscribers (source=didUnsubscribeFrom)
2020-06-27 11:35:35.198001+0100 C19X[12252:5639012] [Transmitter] Automatic beacon code update (lastUpdate=27/06/2020 11:29,elapsed=367.3659030199051)
2020-06-27 11:35:35.198283+0100 C19X[12252:5639012] [Transmitter] updateBeaconCode
2020-06-27 11:35:35.202116+0100 C19X[12252:5639012] [Transmitter] updateBeaconCode successful (code=-766269275189803128,characteristic=0022D481-83FE-1F13-F55D-AA2772040788)`
4. Device returns in range (UUID changed to 31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
`2020-06-27 11:39:37.833988+0100 C19X[12252:5638983] [Transmitter] Subscribe (central=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
2020-06-27 11:39:45.725112+0100 C19X[12252:5639012] [Transmitter] Write (central=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)`
5. Transmitter provides Receiver with scan hint for new UUID
`2020-06-27 11:39:45.727606+0100 C19X[12252:5639012] [Receiver] scan hint found unknown peripheral (source=transmitter|write,peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
2020-06-27 11:39:45.730772+0100 C19X[12252:5639012] [Receiver] scan hint resolved unknown peripheral (source=transmitter|write,peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)`
6. Receiver now able to resume connection 👍
`2020-06-27 11:39:45.731363+0100 C19X[12252:5639012] [Receiver] connect (source=scanHint|central|unknown|.disconnected,peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
2020-06-27 11:39:45.739312+0100 C19X[12252:5638983] [Receiver] didConnect (peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
2020-06-27 11:39:45.739763+0100 C19X[12252:5638983] [Receiver] readCode (source=didConnect,peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
2020-06-27 11:39:45.758552+0100 C19X[12252:5638983] [Receiver] didDiscoverServices (peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9,error=nil)
2020-06-27 11:39:45.759042+0100 C19X[12252:5638983] [Receiver] didDiscoverServices, found beacon service (peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9)
2020-06-27 11:39:45.877650+0100 C19X[12252:5638983] [Receiver] didDiscoverCharacteristicsFor (peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9,error=nil)
2020-06-27 11:39:45.878488+0100 C19X[12252:5638983] [Receiver] didDiscoverCharacteristicsFor, found beacon characteristic (peripheral=31E1087C-C29B-AF7C-D233-CA3DB7329AF9,beaconCode=-4648336143993261413,notifies=true,os=ios)`

SUCCESS AT LAST!

For testing, I have now got a faraday bag, so putting a phone in the bag is equivalent to going out of range. Here are some example test sequences that have worked.

1. Known phone going out of range and back again
    a. Phone A and Phone B starts, detect each other, then both enter B/S state
    b. Phone B is placed in faraday bag, Phone A shows disconnect and enters S state
    c. Phone B taken out of bag, Phone A shows reconnect and return to B state
2. New phone coming in range of Phone in B/S state
    a. Phone A starts, enter B/S state, placed in faraday bag
    b. Phone B starts, enter B/S state
    c. Phone A taken out of bag, Phone B shows discovery and connect
3. New phone coming in range of Phone in T state
    a. Phone A starts, enter B/S state, placed in faraday bag
    b. Phone B starts, use simulate crash to enter T state
    c. Phone A taken out of bag, Phone B shows discovery and connect

I'm going to do more real world tests today, hopefully this really is it 👍

**Comment by UsamaAbdulAziz** [comment]
Excellent observation. That’s definitely the issue. On disconnection, UUID changes, thus not detected. I have tested a lot of cases and till now every one of them works absolutely fine. Amazing and wonderful work. 
This looks like the first complete solution and one of it’s kind. 👍👍👍
Will test it further and update with my findings.

**Comment by c19x** [comment]
That’s great to hear! I’ve been doing lots of tests today and all possible combinations are working well. Shall upload doc soon. The only two situations that are not working are ...

1. Phone A terminates while in Faraday bag. Phone B terminates, then airplane mode On, then Off. Phone A out of Faraday bag. It is not detected. Mitigated by asking user to open app after airplane mode.
2. Phone A terminates while in Faraday bag. Phone B terminates. Wait 30 minutes. Phone A out of Faraday bag, it triggers Phone B to wake up 👍 BUT phone B receiver fails to discover phone A service, so it doesn’t detect beacon. This looks more like a logic issue in my code. Shall investigate. Suspect related to the UUID change problem.

This means going out of range for a short while works but a long time doesn’t at the moment. I am however encouraged by the results and it looks achievable. Your involvement in this is much appreciated and motivating 👍

**Comment by UsamaAbdulAziz** [comment]
For the past few hours, I have testing different scenarios and it works absolutely perfect.
There is only one issue that I observed. Not always but quite a few times. Following are the steps:

1) Installed C19x on Phone A and phone B.
2) Moved both apps to terminated, by simulating a crash.
3) Moved both of them in range, and both Phone A and phone B logged "Contact logs".
4) Switched Bluetooth OFF in Phone A and moved it away from range of Phone B.
5) In Phone A, switched Bluetooth ON, opened C19x and simulated crash.
6) Moved Phone A back in range.
7) Contact start getting logged in Phone B, not in Phone A.

I checked logs from Phone A and I see notify and write logs, and no contact logs.

![Screen Shot 2020-06-27 at 10 05 15 PM](https://user-images.githubusercontent.com/22097793/85930127-548e4e80-b8c2-11ea-9ce6-247d0357ea93.png)

**Comment by UsamaAbdulAziz** [comment]
> That’s great to hear! I’ve been doing lots of tests today and all possible combinations are working well. Shall upload doc soon. The only two situations that are not working are ...
> 
> 1. Phone A terminates while in Faraday bag. Phone B terminates, then airplane mode On, then Off. Phone A out of Faraday bag. It is not detected. Mitigated by asking user to open app after airplane mode.
> 2. Phone A terminates while in Faraday bag. Phone B terminates. Wait 30 minutes. Phone A out of Faraday bag, it triggers Phone B to wake up 👍 BUT phone B receiver fails to discover phone A service, so it doesn’t detect beacon. This looks more like a logic issue in my code. Shall investigate. Suspect related to the UUID change problem.
> 
> This means going out of range for a short while works but a long time doesn’t at the moment. I am however encouraged by the results and it looks achievable. Your involvement in this is much appreciated and motivating 👍

Yes this looks like the issue in handling UUID change on disconnect in ST states.
Your work is unprecedented in this solution to a problem the world is in right now. I am glad that I was of any help in this regard. 👍 👍

**Comment by c19x** [comment]
Thanks for that!!! That is exactly what I needed to resolve the second issue I mentioned in the last comment 🥳. I observed exactly the same thing where the Transmitter on one phone runs in a notify - write loop after both being terminated and out of range for 30 minutes. The other phone is working absolutely fine. This is definitely to do with UUID change. What you’ve just done is given me a quick procedure to replicate the problem, rather then wait 30mins per test. It’s priceless 👏👍 thank you v much. I reckon it’s solvable 🤞

**Comment by UsamaAbdulAziz** [comment]
Perfect!! I will continue with thorough testing. Let me know if I can help with anything else. Kudus to you. 👏 👏

**Comment by c19x** [comment]
Test candidate ready in latest commit on Master branch. Think it is all working 🤞. Have tried every combination and observed UUID change several times being handled gracefully 🥳 Have also repeated your scenario many times, and worked every time so far. Have added Resources/tests/BeaconTests.txt to document tests and results in chronological order. Haven't found a way to trigger UUID change on demand, but running tests repeatedly and toggling bluetooth regularly have generated several examples.

Looking promising, more real world testing tomorrow 👍

**Comment by UsamaAbdulAziz** [comment]
Bravo! Working perfectly fine. I initiated another test and it got me confused. Steps:

1) I put Phone A and Phone B in background while connected and in range. 
2) Contacts are getting logged in both phones. 
3) I move Phone B away from Phone A and put it there, for an hour (just for safe side, might work with less time). 
4) Moved Phone A close to Phone B again.
5) No contacts are getting logged.
6) Checked Console and there is no log with org.c19x. Application somehow looks like in ST states without CoreBluetooth detecting other device. 

I am confused with this scenario as this test should show the same results as the case in which we move both phones away, simulate termination and move them back again. 

One interesting thing here is that I put on a user notification for when transceiver starts with source “BGAppRefreshTask” (i.e. RSSI -10010). I got a few of these notifications in between this one hour period. 🤔

**Comment by c19x** [comment]
I've just seen the same thing too. All scenarios based on simulated crash works, but several real world tests based on moving out of range and waiting for 10+ minutes doesn't work until one of the phones is manually brought back to foreground 🤦🏻‍♂️

My test procedure to replicate this without actually moving away is

1. Phone A and B are connected
2. Phone B placed in faraday bag, wait 10-30 minutes (Phone B UUID = 3BAFAF48-B819-7256-EAC2-EB7DF2FBAA9D)
3. Phone B taken out of faraday bag, not detected by Phone A
4. Phone A to foreground, Phone B immediately detected (Phone B UUID = D7F08A9F-278B-74A2-3503-9F672B64604A)

This is caused by the UUID change. More investigation needed to understand why this is happening. So close but not quite.

I've also seen BGAppRefreshTask running in the background. It starts transmitter (advertise) and receiver (pending scan) but that doesn't seem to do anything. I've disabled it for now to focus on CoreBluetooth debugging.

**Comment by c19x** [comment]
Sharing more info for debugging. I think we are fairly certain root cause is iOS initiated UUID change while Phone is out of range. Here are my observations ...

- `central.scan` only calls `didDiscover` for completely new peripherals
- `central.connect` only calls `didConnect` for peripheral with specific UUID
- If UUID changes while out of range, the pending connect is not satisfied, so no call to `didConnect`
- At the same time `central.scan` considers the peripheral to be known (which is true), albeit with a new UUID, so no call to `didDiscover` either 😓

I imagine a call to `central.retrieveConnectedPeripherals` *should* list the peripheral with new UUID ... but something needs to trigger that to happen. `BGAppRefresh` has only fired while out of range, but not after the phone has gone back in range (why?), so can't depend on that to call retrieve either. `LocationManager` is more reliable for firing events, but that never used to do anything to CoreBluetooth calls either (why? wrong thread?).

At least the problem can now be consistently reproduced for investigation. I've just ran the following test ...

1. Phone A and B are connected
2. Phone B goes out of range for 20 minutes
3. Phone B returns in range, not detected by Phone A
4. New Android Phone C comes in range, detected by both Phone A and B, but A and B still do not detect each other
5. Phone A goes foreground, immediately detected by Phone B (UUID has changed)

This suggests its not about CoreBluetooth FBST state, and more about `central.scan` not reporting known peripheral even though when its UUID has changed 🤔 Tricky.

**Comment by c19x** [comment]
Reading your test sequence again. I wonder if we can distill it down to ...

1. Phone A and B connected
2. Phone A Bluetooth OFF
3. Phone B in faraday bag
4. Phone A Bluetooth ON, simulate crash
5. Phone B out of faraday bag

... transmitter on Phone A in notify - write loop. Thiscaused by Phone A UUID change.

**Comment by UsamaAbdulAziz** [comment]
> Sharing more info for debugging. I think we are fairly certain root cause is iOS initiated UUID change while Phone is out of range. Here are my observations ...
> 
> 
> 
> - `central.scan` only calls `didDiscover` for completely new peripherals
> 
> - `central.connect` only calls `didConnect` for peripheral with specific UUID
> 
> - If UUID changes while out of range, the pending connect is not satisfied, so no call to `didConnect`
> 
> - At the same time `central.scan` considers the peripheral to be known (which is true), albeit with a new UUID, so no call to `didDiscover` either 😓
> 
> 
> 
> I imagine a call to `central.retrieveConnectedPeripherals` *should* list the peripheral with new UUID ... but something needs to trigger that to happen. `BGAppRefresh` has only fired while out of range, but not after the phone has gone back in range (why?), so can't depend on that to call retrieve either. `LocationManager` is more reliable for firing events, but that never used to do anything to CoreBluetooth calls either (why? wrong thread?).
> 
> 
> 
> At least the problem can now be consistently reproduced for investigation. I've just ran the following test ...
> 
> 
> 
> 1. Phone A and B are connected
> 
> 2. Phone B goes out of range for 20 minutes
> 
> 3. Phone B returns in range, not detected by Phone A
> 
> 4. New Android Phone C comes in range, detected by both Phone A and B, but A and B still do not detect each other
> 
> 5. Phone A goes foreground, immediately detected by Phone B (UUID has changed)
> 
> 
> 
> This suggests its not about CoreBluetooth FBST state, and more about `central.scan` not reporting known peripheral even though when its UUID has changed 🤔 Tricky.

You are right. Have been testing a few things but I didn’t get any success at all. Well I am not hopeless. Will let you know if I find anything.

**Comment by c19x** [comment]
Just hit the wrong button there 😑 didn’t mean to close issue.

After lots of tests today here are the findings ...

1. I’ve used LocationManager to enable on demand calls to retrieveConnectedPeripherals, but sadly it doesn’t list the peripheral with changed UUID.
2. Connect definitely does not handle change UUID, so didConnect is not called.
3. Scan definitely does not handle change UUID either, it must think peripheral is known and won’t call didDiscover until app returns to foreground on one of the devices.
4. Change UUID was handled correctly ONCE in the test, due to the most bizarre behaviour where a pending characteristic subscribe sneaked in after disconnect (!) and while Bluetooth was OFF (!!!). I’ve recorded log for analysis.
5. Some subscribe events seem to continue even with Bluetooth OFF. I’m baffled.

Try this for total weirdness ...

1. Phone A and B connected
2. Phone A in faraday bag, Phone B shows disconnect
3. Phone B Bluetooth OFF, in suspended state
4. Phone A out of faraday bag, Phone B shows transmitter events 😮 triggered by Phone A (how?)
5. Phone B Bluetooth ON, all works normally.

This weird behaviour is what made the change UUID work that once. Give it a go, super strange.

**Comment by UsamaAbdulAziz** [comment]
I have observed this issue with BLE events while Bluetooth is OFF but ignored it altogether. Now idea what sorcery was that  🤨
No luck from my side. Will try some new ideas as well.🤞

**Comment by c19x** [comment]
Here is the full log for that one instance where a UUID change was survived, may help us to find a way out ...

LOG : UUID CHANGE SURVIVED 
==========================================================================================================================================================
=== Phone A and B connected
debug	15:28:23.346220+0100	C19X	Receiver	readRSSI (source=didUpdateValueFor,peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219)
debug	15:28:23.349169+0100	C19X	Receiver	didReadRSSI (peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219,rssi=-57,error=nil)
debug	15:28:23.350653+0100	C19X	Database	insert (time=28/06/2020 15:28,code=4276855838531190253,rssi=-57)
debug	15:28:23.358923+0100	C19X	Receiver	scan (source=scheduleScan|didWriteValueFor,statistics={count=7,mean=8.01845371723175,sd=6.564739438710513,min=0.09487795829772949,max=18.104115962982178,period=56.12917602062225})
default	15:28:23.359813+0100	C19X	Receiver	Beacon state report ========
default	15:28:23.360610+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connected)
debug	15:28:23.369347+0100	C19X	Controller	Contact logged (code=4276855838531190253,rssi=-57,count=329,timestamp=28/06/2020 15:28)
debug	15:28:23.370385+0100	C19X	ViewController	transceiver did detect contact (timestamp=28/06/2020 15:28)
debug	15:28:23.370882+0100	C19X	ViewController	updateViewData (status=false,contacts=true,advice=false,suppressNotification=false)
debug	15:28:23.371376+0100	C19X	Receiver	Detected beacon (source=didReadRSSI,peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219,code=4276855838531190253,rssi=-57,statistics={count=1,mean=8.686784982681274,sd=-,min=8.686784982681274,max=8.686784982681274,period=8.686784982681274})
debug	15:28:23.389976+0100	C19X	Receiver	wakeTransmitter (source=scan|ios,peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219)
debug	15:28:23.435225+0100	C19X	Receiver	didWriteValueFor (peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219,error=nil)
=== Phone A in faraday bag

debug	15:28:24.662246+0100	C19X	Receiver	didDisconnectPeripheral (peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219,error=Optional(Error Domain=CBErrorDomain Code=6 "The connection has timed out unexpectedly." UserInfo={NSLocalizedDescription=The connection has timed out unexpectedly.}))
debug	15:28:24.662889+0100	C19X	Receiver	connect (source=didDisconnectPeripheral,peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219)
debug	15:28:24.663664+0100	C19X	Transmitter	Unsubscribe (central=6FE588B2-161F-4CED-667A-FC8F2AD62219)
debug	15:28:32.757138+0100	C19X	Transmitter	Notify subscribers (source=didUnsubscribeFrom)
debug	15:28:32.757862+0100	C19X	Receiver	scan (source=scheduleScan|connect,statistics={count=8,mean=8.190945118665695,sd=6.097317996684063,min=0.09487795829772949,max=18.104115962982178,period=65.52756094932556})
default	15:28:32.758791+0100	C19X	Receiver	Beacon state report ========
default	15:28:32.772999+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
debug	15:28:47.365664+0100	C19X	LocationManager	didExitRegion
debug	15:28:47.366739+0100	C19X	LocationManager	locationUpdate (type=region)
debug	15:28:47.367144+0100	C19X	Receiver	scan (source=locationManager,statistics={count=9,mean=8.904223892423841,sd=6.0917177700913525,min=0.09487795829772949,max=18.104115962982178,period=80.13801503181458})
default	15:28:47.367609+0100	C19X	Receiver	Beacon state report ========
default	15:28:47.368401+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
debug	15:28:47.370388+0100	C19X	Transceiver	Beacon state report (subscribers) ========
debug	15:28:47.455692+0100	C19X	LocationManager	didUpdateLocations
debug	15:28:47.455823+0100	C19X	LocationManager	locationUpdate (type=location)
debug	15:28:47.456415+0100	C19X	Receiver	scan (source=locationManager,statistics={count=10,mean=8.022420597076415,sd=6.3844789030836,min=0.08619093894958496,max=18.104115962982178,period=80.22420597076416})
default	15:28:47.456593+0100	C19X	Receiver	Beacon state report ========
default	15:28:47.456705+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
debug	15:28:47.468954+0100	C19X	Transceiver	Beacon state report (subscribers) ========
debug	15:28:47.476335+0100	C19X	LocationManager	startMonitoring (region=CLCircularRegion (identifier:'Circle', center:<+51.50998000,-0.13370000>, radius:10.00m))
debug	15:29:38.681339+0100	C19X	Transmitter	Update state (state=.poweredOff)
debug	15:29:38.685971+0100	C19X	Receiver	State updated (toState=.poweredOff)
debug	15:29:38.686145+0100	C19X	Controller	Bluetooth state updated (state=.poweredOff)
debug	15:29:38.686408+0100	C19X	ViewController	transceiver did update state (state=.poweredOff)
debug	15:29:38.693762+0100	C19X	ViewController	notification (method=background,title=Contact Tracing Disabled,body=Turn ON Bluetooth to resume.)
=== Phone B Bluetooth power OFF


=== Phone A out of faraday bag, subscription request under new UUID detected while Bluetooth is OFF !!!!!
debug	15:29:45.127459+0100	C19X	Transmitter	Subscribe (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:29:45.127765+0100	C19X	Receiver	scan hint found unknown peripheral (source=transmitter|didSubscribeTo,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
fault	15:29:45.129446+0100	C19X	Receiver	scan hint cannot resolve unknown peripheral (source=transmitter|didSubscribeTo,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:29:53.047529+0100	C19X	Transmitter	Write (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:29:53.048414+0100	C19X	Receiver	scan hint found unknown peripheral (source=transmitter|didReceiveWrite,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
fault	15:29:53.050416+0100	C19X	Receiver	scan hint cannot resolve unknown peripheral (source=transmitter|didReceiveWrite,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:01.143002+0100	C19X	Transmitter	Notify subscribers (source=didReceiveWrite)
debug	15:30:01.143934+0100	C19X	Transmitter	Automatic beacon code update (lastUpdate=28/06/2020 15:27,elapsed=145.89844691753387)
debug	15:30:01.144095+0100	C19X	Transmitter	updateBeaconCode
fault	15:30:01.144506+0100	C19X	Transmitter	updateBeaconCode denied, bluetooth is not on
debug	15:30:06.162201+0100	C19X	Transmitter	Unsubscribe (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:06.162470+0100	C19X	Receiver	scan hint found unknown peripheral (source=transmitter|didUnsubscribeFrom,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
fault	15:30:06.164773+0100	C19X	Receiver	scan hint cannot resolve unknown peripheral (source=transmitter|didUnsubscribeFrom,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:14.357383+0100	C19X	Transmitter	Automatic beacon code update (lastUpdate=28/06/2020 15:27,elapsed=159.04781794548035)
debug	15:30:14.357434+0100	C19X	Transmitter	updateBeaconCode
fault	15:30:14.357662+0100	C19X	Transmitter	updateBeaconCode denied, bluetooth is not on
debug	15:30:14.357841+0100	C19X	Transmitter	Notify subscribers (source=didUnsubscribeFrom)

=== Phone A back in faraday bag, Phone B Bluetooth power ON
debug	15:30:18.970437+0100	C19X	Transmitter	Update state (state=.poweredOn)
debug	15:30:18.970541+0100	C19X	Transmitter	start (source=didUpdateState|powerOn)
debug	15:30:18.970726+0100	C19X	Transmitter	starting advert for existing characteristic
debug	15:30:18.971455+0100	C19X	Transmitter	start successful, for existing beacon code (source=didUpdateState|powerOn,code=7207958562117079220)
debug	15:30:19.006721+0100	C19X	Receiver	State updated (toState=.poweredOn)
debug	15:30:19.006875+0100	C19X	Receiver	scan (source=updateState,statistics={count=11,mean=15.61200154911388,sd=25.89024030557377,min=0.08619093894958496,max=91.50781106948853,period=171.73201704025269})
default	15:30:19.007196+0100	C19X	Receiver	Beacon state report ========
default	15:30:19.008968+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.disconnected)
debug	15:30:19.010509+0100	C19X	Controller	Bluetooth state updated (state=.poweredOn)
debug	15:30:19.011265+0100	C19X	ViewController	transceiver did update state (state=.poweredOn)
debug	15:30:19.062510+0100	C19X	Receiver	connect (source=scan|ios|.disconnected,peripheral=6FE588B2-161F-4CED-667A-FC8F2AD62219)
debug	15:30:27.045341+0100	C19X	Transmitter	Notify subscribers (source=start|didUpdateState|powerOn)
debug	15:30:27.053392+0100	C19X	Transmitter	Automatic beacon code update (lastUpdate=28/06/2020 15:27,elapsed=171.8009089231491)
debug	15:30:27.053507+0100	C19X	Receiver	scan (source=scheduleScan|connect,statistics={count=12,mean=14.984684000412624,sd=24.7808412714256,min=0.08619093894958496,max=91.50781106948853,period=179.81620800495148})
default	15:30:27.053566+0100	C19X	Receiver	Beacon state report ========
debug	15:30:27.053845+0100	C19X	Transmitter	updateBeaconCode
default	15:30:27.054188+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
debug	15:30:27.055946+0100	C19X	Transmitter	updateBeaconCode successful (code=-1068332789430119520,characteristic=0022D481-83FE-1F13-F12C-84FD001D93A0)
debug	15:30:37.362052+0100	C19X	LocationManager	didExitRegion
debug	15:30:37.365562+0100	C19X	LocationManager	locationUpdate (type=region)
debug	15:30:37.365909+0100	C19X	Receiver	scan (source=locationManager,statistics={count=13,mean=14.62588500059568,sd=23.761092083469773,min=0.08619093894958496,max=91.50781106948853,period=190.13650500774384})
default	15:30:37.366075+0100	C19X	Receiver	Beacon state report ========
default	15:30:37.366591+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
debug	15:30:37.367752+0100	C19X	Transceiver	Beacon state report (subscribers) ========
debug	15:30:37.457904+0100	C19X	LocationManager	didUpdateLocations
debug	15:30:37.458093+0100	C19X	LocationManager	locationUpdate (type=location)
debug	15:30:37.458611+0100	C19X	Receiver	scan (source=locationManager,statistics={count=14,mean=13.587808217321124,sd=23.156985628598893,min=0.08619093894958496,max=91.50781106948853,period=190.22931504249573})
default	15:30:37.458976+0100	C19X	Receiver	Beacon state report ========
default	15:30:37.460531+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
debug	15:30:37.467752+0100	C19X	Transceiver	Beacon state report (subscribers) ========
debug	15:30:37.474234+0100	C19X	LocationManager	startMonitoring (region=CLCircularRegion (identifier:'Circle', center:<-26.20410280,+28.04730510>, radius:10.00m))

=== Phone A out of faraday bag, pending subscription and resolved peripheral under new UUID, back to normal running!
debug	15:30:47.828876+0100	C19X	Transmitter	Subscribe (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.828984+0100	C19X	Receiver	scan hint found unknown peripheral (source=transmitter|didSubscribeTo,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.829178+0100	C19X	Receiver	scan hint resolved unknown peripheral (source=transmitter|didSubscribeTo,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.829226+0100	C19X	Receiver	connect (source=scanHint|central|unknown|.disconnected,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.830425+0100	C19X	Receiver	didConnect (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.830573+0100	C19X	Receiver	readCode (source=didConnect,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.833264+0100	C19X	Receiver	didDiscoverServices (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)
debug	15:30:47.833372+0100	C19X	Receiver	didDiscoverServices, found beacon service (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:47.949784+0100	C19X	Receiver	didDiscoverCharacteristicsFor (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)
debug	15:30:47.950138+0100	C19X	Receiver	didDiscoverCharacteristicsFor, found beacon characteristic (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,beaconCode=4276855838531190253,notifies=true,os=ios)
debug	15:30:55.781625+0100	C19X	Transmitter	Write (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:30:56.040052+0100	C19X	Receiver	scan (source=scheduleScan|didDiscoverCharacteristicsFor,statistics={count=15,mean=13.920727801322938,sd=22.351850720142153,min=0.08619093894958496,max=91.50781106948853,period=208.81091701984406})
default	15:30:56.040975+0100	C19X	Receiver	Beacon state report ========
default	15:30:56.041553+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
default	15:30:56.042081+0100	C19X	Receiver	Beacon state (uuid=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,state=.connected)
debug	15:30:56.069967+0100	C19X	Receiver	scan found duplicate peripheral (code=4276855838531190253,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,duplicateOf=6FE588B2-161F-4CED-667A-FC8F2AD62219)
debug	15:30:56.112004+0100	C19X	Receiver	didUpdateValueFor (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)
debug	15:30:56.112060+0100	C19X	Receiver	readRSSI (source=didUpdateValueFor,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:03.879128+0100	C19X	Transmitter	Notify subscribers (source=didReceiveWrite)
debug	15:31:03.940711+0100	C19X	Transmitter	Write (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:03.942424+0100	C19X	Receiver	scan hint found unknown peripheral (source=transmitter|didReceiveWrite,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:03.944504+0100	C19X	Receiver	scan hint resolved unknown peripheral (source=transmitter|didReceiveWrite,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:03.945137+0100	C19X	Receiver	connect (source=scanHint|central|unknown|.disconnected,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:03.952637+0100	C19X	Receiver	didConnect (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:03.953290+0100	C19X	Receiver	readCode (source=didConnect,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:04.061424+0100	C19X	Receiver	didDiscoverServices (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)
debug	15:31:04.061753+0100	C19X	Receiver	didDiscoverServices, found beacon service (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:04.184408+0100	C19X	Receiver	didDiscoverCharacteristicsFor (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)
debug	15:31:04.185122+0100	C19X	Receiver	didDiscoverCharacteristicsFor, found beacon characteristic (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,beaconCode=-1051892813811875473,notifies=true,os=ios)
debug	15:31:11.955889+0100	C19X	Transmitter	Notify subscribers (source=didReceiveWrite)
debug	15:31:12.039786+0100	C19X	Transmitter	Write (central=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:12.288373+0100	C19X	Receiver	scan (source=scheduleScan|didDiscoverCharacteristicsFor,statistics={count=16,mean=14.066129311919214,sd=21.60177031702935,min=0.08619093894958496,max=91.50781106948853,period=225.0580689907074})
default	15:31:12.288912+0100	C19X	Receiver	Beacon state report ========
default	15:31:12.289969+0100	C19X	Receiver	Beacon state (uuid=6FE588B2-161F-4CED-667A-FC8F2AD62219,state=.connecting)
default	15:31:12.290864+0100	C19X	Receiver	Beacon state (uuid=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,state=.connected)
debug	15:31:12.322412+0100	C19X	Receiver	wakeTransmitter (source=scan|ios,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:12.341161+0100	C19X	Receiver	didUpdateValueFor (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)
debug	15:31:12.341839+0100	C19X	Receiver	readRSSI (source=didUpdateValueFor,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF)
debug	15:31:12.345283+0100	C19X	Receiver	didReadRSSI (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,rssi=-39,error=nil)
debug	15:31:12.346934+0100	C19X	Database	insert (time=28/06/2020 15:31,code=-1051892813811875473,rssi=-39)
debug	15:31:12.367258+0100	C19X	Controller	Contact logged (code=-1051892813811875473,rssi=-39,count=330,timestamp=28/06/2020 15:31)
debug	15:31:12.368940+0100	C19X	ViewController	transceiver did detect contact (timestamp=28/06/2020 15:31)
debug	15:31:12.369965+0100	C19X	ViewController	updateViewData (status=false,contacts=true,advice=false,suppressNotification=false)
debug	15:31:12.371018+0100	C19X	Receiver	Detected beacon (source=didReadRSSI,peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,code=-1051892813811875473,rssi=-39,statistics={count=0,mean=-,sd=-,min=-,max=-,period=0.0})
debug	15:31:12.371761+0100	C19X	Receiver	didWriteValueFor (peripheral=AD6EB0BF-9159-3541-C7B0-2BB231483AEF,error=nil)

**Comment by c19x** [comment]
Just pushed new branch ReconnectOutOfRange to share experimentation code (see TestTransceiver.swift for vastly simplified BLE code for testing options and tracing what is happening). Please use simulated location update to trigger scan on-demand if required, but it should run in a basic loop on discovery. Key ideas under test are:

1. Removal of strong references to peripheral on disconnect
2. Minimise reliance on strong references to peripheral and refresh where possible in delegate calls
3. Use retrievePeripherals and retrieveConnectedPeripherals as primary search to support scan

So far, I am seeing that certain calls do survive UUID change, in particular setNotifyValue and didDiscover ...

Latest test shows ...
1. Phone A and B connected
2. Phone A discoverService call to Phone B
3. Phone B placed in faraday bag during discoverService call
4. Wait
5. Phone B out of faraday bag (new UUID)
6. Phone A resumes with didSubscribeTo or didDiscover delegate call

Looks like discoverService -> discoverCharacteristic -> setNotifyValue happened somewhere else internally in one instance, and discoverService -> didDisconnectPeripheral -> didUnsubscribeFrom happened somewhere else internally in another instance.

This offers a bit of hope.

**Comment by UsamaAbdulAziz** [comment]
> Just pushed new branch ReconnectOutOfRange to share experimentation code (see TestTransceiver.swift for vastly simplified BLE code for testing options and tracing what is happening). Please use simulated location update to trigger scan on-demand if required, but it should run in a basic loop on discovery. Key ideas under test are:
> 
> 
> 
> 1. Removal of strong references to peripheral on disconnect
> 
> 2. Minimise reliance on strong references to peripheral and refresh where possible in delegate calls
> 
> 3. Use retrievePeripherals and retrieveConnectedPeripherals as primary search to support scan
> 
> 
> 
> So far, I am seeing that certain calls do survive UUID change, in particular setNotifyValue and didDiscover ...
> 
> 
> 
> Latest test shows ...
> 
> 1. Phone A and B connected
> 
> 2. Phone A discoverService call to Phone B
> 
> 3. Phone B placed in faraday bag during discoverService call
> 
> 4. Wait
> 
> 5. Phone B out of faraday bag (new UUID)
> 
> 6. Phone A resumes with didSubscribeTo or didDiscover delegate call
> 
> 
> 
> Looks like discoverService -> discoverCharacteristic -> setNotifyValue happened somewhere else internally in one instance, and discoverService -> didDisconnectPeripheral -> didUnsubscribeFrom happened somewhere else internally in another instance.
> 
> 
> 
> This offers a bit of hope.

This sure is a step towards the solution. I will check out the branch and will report if I find a way to fix this.🤞

**Comment by UsamaAbdulAziz** [comment]
I am away from my system, but I need your opinion on this idea. What do you say if:

1) We connect peripheral only if the RSSI is greater than a significant value (say -90). 
2) Make readRSSI call more frequent to observe if it goes less than this RSSI. 
3) As soon as a peripheral has gone below this defined RSSI we disconnect it from central, as our app will be in background state with our methods still functioning before getting terminated. 

Would central still consider this peripheral as known device or will it be a new device for it?

**Comment by c19x** [comment]
I think your idea is sound, but I expect CoreBluetooth will give us trouble due to UUID change and also background scan not calling didDiscover. Consider this scenario ...

1. Phone A is scanning. Phone B is unknown. A detects B (didDiscover called for first time). A does not connect due to low RSSI.
2. Phone A can only get RSSI update again by calling connect, readRSSI, disconnect. Even if we can trigger a scanForPeripheral call, didDiscover won’t be called a second time ☹️ So we can’t avoid a connect, unless a readRSSI call can be made without connect? Shall investigate.

What I have discovered today is that there are two kinds of UUID change and a reliable way to trigger and also handle one case.

The case we can handle is ...

1. Phone A and B connected
2. Phone A airplane mode ON, B registers didDisconnectPeripheral and didUnsubscribe
3. Phone A airplane mode OFF, B registers didSubscribe by A with new UUID and resumes connection indefinitely.

Take a look at the last test log in file Resources/tests/TestProcedures.txt 

The case we can’t still handle yet is UUID change that occurs within iOS after a period of inactivity, following didDisconnect. See test 15:40:35 in test log. However, if the disconnect happens to occur during a discoverService call, such that didDisconnect and didUnsubscribe are not called, then it will survive the UUID change and come back via didSubscribe. I’ve seen that a few times today by chance and have recorded log as evidence for analysis.

If didDiscover is called for all iOS devices following a scanForPeripherals call (like it does with Android devices), this will be straightforward. I have no idea why someone decided it was a good idea to only report iOS devices on first discovery only, and then change the UUID later (to comply with Bluetooth MAC address change spec?) such that the reported device cannot be retrieved. I guess it is an unusual use case. The fact that an interrupted discoverService call is resumable even after UUID change is interesting 🤔 unless it was just lucky (that many times?).

I think the kind of UUID change that we can’t handle yet occurs when the peripheral is no longer connected at the os level (not in retrieveConnectedPeripherals). I bet it has two identifiers, one consistent id for internal device caching and another that we are seeing at app level that changes as per Bluetooth spec. The scanForPeripheral call uses the former for de-duplication and CoreBluetooth uses it for state restoration and resuming pending actions. We get the latter as reference for peripheral retrieval but there is no callback for id change. The ground truth is the reference to the internal CBPeripheral object that I’m trying to extract from the delegate callbacks. We need to find a pending call that fires reliably to get a callback 🧐🔍

**Comment by UsamaAbdulAziz** [comment]
Tried many options and procedures but couldn’t handle this UUID change (by OS). As you said, I have been able to get success a few times, but not majority of times. Let’s keep digging.. 😓

**Comment by c19x** [comment]
It sure is a tough one, but it’s worth persevering, especially as G+A api is being rejected in most countries for various reasons.

I have managed to make a bit of progress today, but confused by how it is happening. Here is a repeatable scenario ...

1. A and B connected
2. A, airplane mode ON, B registers disconnect and unsubscribe, B is idle on scheduled scan call
3. Wait 1 minute
4. A, airplane mode OFF, B registers didSubscribeTo (new UUID), calls updateValue, then resumes scheduled scan call and resumes connection as normal on new UUID

This is strange because the first scheduled scan call immediately schedules a second scan call after at least 8 seconds, but in this case that second scan call appears to be held dormant until A is detected again by B. I can’t get it to work if the delay is too long at the moment but at least we now have a repeatable example that is handled correctly, albeit for a short delay. Shall push branch soon, so you can have a look too 👍

**Comment by UsamaAbdulAziz** [comment]
Right. Let’s see if this helps in getting the solution. I’ll look into the code as soon as you push the branch. 👍👍

**Comment by c19x** [comment]
Just pushed the branch, include work in progress. Please check out commit "1879dfa" tagged "ChangeUUID_AirplaneModeSurvival", it shows exactly which aspects of the code enables UUID change survival (for a little while). I think its to do with a de-alloc delay or glitch, rather than a CoreBluetooth feature.

I've been trying another idea, to use rotating Service UUID and rotating scan UUIDs to see if that triggers didDiscover, but no luck. I've got another idea to try out, this one takes a bit longer. Shall keep you posted 👍

**Comment by c19x** [comment]
Hello! Just created new repo C19X-iOS-BLE with bare minimum BLE code for investigating this out of range resume after long periods while in B/S/T state problem and documenting findings. I've brought this project back to most complete state, i.e. all functional except this specific problem, as I was in danger of breaking intricate functional code while experimenting with lots of different ideas. The bare minimum code is also easier to trace than the full app. Please see README.MD for ideas that I've tried so far.

**Comment by UsamaAbdulAziz** [comment]
Great. I will check out that repo to resume experimentation work. Seems like this discrepancy in BLE delegates is deliberately programmed by Apple for some God knows what reason. Will update if I get any success. 👍

**Comment by UsamaAbdulAziz** [comment]
Just tested your minimalistic BLE repo. I want to share some observations here:

1) Installed app on phone A and terminated it. Enabled airplane mode just to be sure. 
2) Installed app on phone B and locked the screen, changing its state to background. 
3) Moved A out of range, disabled airplane mode and launched app. 
4) Both phones showed log (init, init.delegate, scan, startAdvertising, didStartAdvertising). 
5) After 30 mins delay, returned A back within range of B and there was no change in logs. It remained same as mentioned in point#4. 
6) Restarted A and as soon as it got restarted, same logs reappeared and then progressed on to didDiscover and eventually logged contact registration as well. 

This makes me wonder if this problem that we are trying to solve, happening with only known devices. Or is there any other perspective to it as well.

**Comment by c19x** [comment]
Wow, thanks for testing the code so thoroughly 👍👍👍  and documenting your findings 👏  What interesting but slightly worrying observations. That suggests if phone A starts, detects nothing, enters and stays in S/T state for over 30mins, then other phones in B/S/T state won’t ever detect it 😬 I hope that is not the case. I shall go test it too. I’ve opened a code-level support ticket with Apple (see latest commit). Looking forward to seeing what they come back with. Shall share responses as they become available. Good work! Let’s keep chipping at it 🤞

**Comment by c19x** [comment]
I've just repeated your test procedure and you are entirely correct. They don't detect each other. O dear 😐. This is looking like a bigger problem than just UUID change. HOWEVER, I think you've just directed us towards the exact cause of the problem and I might even have a workable solution 🥳🥳🤝🤞

To carry on with your experiment ...

While the two iPhones (A and B) are in this non-detected state, I've modified the Android app to make it a BLE sniffer (receive only) on Android phone C. At this point, A and B have not detected each other. I placed B into a faraday bag first, then started the sniffer app on C. I'm pleasantly surprised to find that at least iOS is still advertising the service and characteristic on phone A. Here is the log of C:

- org.c19x.beacon.ConcreteReceiver: Scan started
- org.c19x.beacon.ConcreteReceiver: Processing beacon (uuid=6A:1D:C8:AE:A3:DA,type=iosBackground,code=NULL,rssi=RSSI{value=-47})
- org.c19x.beacon.ConcreteReceiver: GATT client connection state change (device=6A:1D:C8:AE:A3:DA,status=0,newState=STATE_CONNECTED)
- org.c19x.beacon.ConcreteReceiver: GATT client discovered services (device=6A:1D:C8:AE:A3:DA,status=0)
- org.c19x.beacon.ConcreteReceiver: GATT client discovered C19X service (device=6A:1D:C8:AE:A3:DA,transmitterBeaconCode=BeaconCode{value=1})
- org.c19x.beacon.ConcreteReceiver: Detected beacon (beaconCode=BeaconCode{value=1},rssi=RSSI{value=-47})
- org.c19x.beacon.ConcreteReceiver: GATT client connection state change (device=6A:1D:C8:AE:A3:DA,status=0,newState=STATE_DISCONNECTED)
- org.c19x.beacon.ConcreteReceiver: GATT client closed (device=6A:1D:C8:AE:A3:DA)

This suggest the problem we are facing is on the receiver (CBCentralManager) side. Phew 😓! It would have been worse in some ways if it's not even advertising. To test this, I've modified the Android app to make it a BLE advert (transmit only). I stopped the sniffer app on C, placed C (android) by B (iOS) and started the advert app on C. Here is the log on C showing the advert was started.

- org.c19x.beacon.ConcreteTransmitter: setGattService (beaconCode=BeaconCode{value=2},serviceUUID=0022d481-83fe-1f13-0000-000000000000,characteristicUUID=0022d481-83fe-1f13-0000-000000000002)
- org.c19x.beacon.ConcreteTransmitter: transmitting (code=BeaconCode{value=-8724774741551609969})
- org.c19x.beacon.ConcreteTransmitter: Advert started (settings=Settings [mAdvertiseMode=0, mAdvertiseTxPowerLevel=1, mAdvertiseConnectable=true, mAdvertiseTimeoutMillis=0])
 
After 55 seconds, the iOS side discovered the Android advert and started the connect -> disconnect loop. Here is the log on A (iOS).

- 2020-07-05 09:27:38 +0000 [Beacon] [CentralManager(c19x)] : didDiscover (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.disconnected>) -> connect
- 2020-07-05 09:27:38 +0000 [Beacon] [CentralManager(c19x)] : connect (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.disconnected>) -> register, didConnect|didFailToConnect
- 2020-07-05 09:27:38 +0000 [Beacon] [CentralManager(c19x)] : register (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.disconnected>)
- 2020-07-05 09:27:42 +0000 [Beacon] [CentralManager(c19x)] : didConnect (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.connected>) -> didReadRSSI
- 2020-07-05 09:27:42 +0000 [Beacon] [CentralManager(c19x)] : didReadRSSI (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.connected>,rssi=-37) -> disconnect
- 2020-07-05 09:27:42 +0000 [Beacon] [CentralManager(c19x)] : disconnect (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.connected>) -> didDisconnectPeripheral
- 2020-07-05 09:27:42 +0000 [Beacon] [CentralManager(c19x)] : didDisconnectPeripheral (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.disconnected>) -> connect
- 2020-07-05 09:27:42 +0000 [Beacon] [CentralManager(c19x)] : connect (<CBPeripheral-8000:uuid=321EE513-4BB4-4089-6A36-14235EA12738,state=.disconnected>) -> register, didConnect|didFailToConnect

So far so good, demonstrating that iOS background advertising is working and iOS background scan for Android is also working. I let it run for 5 minutes, then stopped the Android advert and the iOS side has stopped connecting as expected.

All this time, the other iOS device B has been sitting in the faraday bag. Given the BLE scan on A (iOS) is clearly functional in some way (detected C), I brought B back out of the faraday bag to see what happens. Surprisingly, A and B are next to each other and B was not detected by A, even though the BLE scan on A was recently powered up 🧐

To complete the test, I have A and B by each other and started the advert on C (android). As before, A (iOS) detects C after about 30 seconds and starts the connect -> disconnect loop. This tells us the BLE receiver on A is powered up and working. However, B is still not detected by A, even in this state. As a confirmation, I switch C (android) back to the sniffer app and it definitely shows both A and B are advertising ...

- org.c19x.beacon.ConcreteReceiver: GATT client discovered C19X service (device=4E:68:F9:CF:2D:56)
- org.c19x.beacon.ConcreteReceiver: GATT client discovered C19X service (device=78:07:ED:BE:46:6F)

All this is telling me that the iOS BLE stack is definitely powered up and running on both A and B. The fact that they are not detecting each other but detectable by Android is therefore caused by something else ... UUID in overflow area which is unique to iOS.

I've started a new branch (detect_iosBackgroundAdvert) which includes a LocationManager that starts beacon ranging in the background. The idea is that UUID data in the overflow area is only delivered while iOS is in background mode when it is ranging for iBeacons and while the screen is on (not necessarily app in foreground, just screen on). The latter can be achieved and automated with local notifications if this works.

SO ... I placed B back in the faraday bag, then started the updated app on A with beacon ranging code running. State restoration brought A back to the still-not-detecting B state (handy, but somewhat ironic), the app was backgrounded and I locked the phone to turn screen off. I took B back out of the faraday bag, waited for a while and nothing happened ... until I touched the screen of A, it lit up ... and detected B and started the connect -> disconnect loop 🥳 problem partially solved.

Next step is to repeat this whole situation again to make sure I didn't do anything wrong to confirm the findings. If it really works, then I shall automate the screen on at regular intervals to see if beacon ranging solves the problem.

Here is the test procedure ...
1. Delete app on A and B, airplane mode ON for both.
2. Airplane mode OFF for A, install app, background, confirm advertising in Xcode log, place in faraday bag.
3. Airplane mode OFF for B, install app, background, keep connected to Xcode, confirm advertising in log.
4. A and B have not detected each other at this point.
5. Wait 30 minutes. Ensure screen is off on B. Take A out of faraday bag. Expect no detection.
6. Wait 3 minutes. Expect no detection. Touch screen of B to activate screen. Expect detection of A and connect -> disconnect loop to run indefinitely.

I shall run this test at the next opportunity and share the results 🤞

**Comment by UsamaAbdulAziz** [comment]
I just tested out your findings and found it to be accurate. I also tested out the new implementation with beacon ranging awakening the app and keeping it in background. It’s a good idea 👍👍
But don’t you think that keeping screen ON with the help of local notifications might irritate the user? Or is there a smart way we can handle it? I’ll post here if I have any idea to achieve this. 
Brilliant work there 🤟🤟

**Comment by c19x** [comment]
Hello! Thanks for testing the idea 👍. I’ve been repeating the test and it does seem to be working well and reliably 🥳 You are right, keeping the screen on a lot would be irritating and also drain the battery. In our case, I think we only need the screen on v briefly once every 5-10 minutes, and the notification can be used to provide useful info (like number of recent contacts and whether it’s active for reassurance). It doesn’t require any user interaction. I’ve implemented it on mine and I don’t notice anything as I have a case, so the screen is normally covered. Yes, please do experiment with ideas for lighting up the screen 👍👍 with a bit of luck, we’ve cracked this past problem 🤝

**Comment by c19x** [comment]
It is all working! 🥳👍 I have introduced the fix to the C19X-iOS code. Just done many tests and it's worked well to discover new peripheral that has been idle for a long time, and also resume connection for a known peripheral after being out of range for >20 mins. I think we finally have a complete solution🤞.

The plan now is to do more real world and faraday bag testing to ensure it works reliably and consistently, and also put all the lessons learnt from C19X-iOS-BLE back into the C19X-iOS app (e.g. improving logging, tidy code, simplify asyncs, remove redundant processes like BGAppRefresh) and also start rolling back the deployment target towards iOS 10 (or even 9.3 if possible, but might not be worth the effort).

On the usability side, I have been thinking about your comment regarding screen ON and repeating notifications. I think we can make that a feature like a micro-GUI that delivers useful info and remind people of good behaviour based on context, such that people rarely need to actually open the app, apart from submitting status update. More thoughts required, but I'm hoping we now have a good foundation for proximity based disease control.

BTW, you may not have noticed, but there is already a feature for setting a device specific message on the server side that is pulled by the device and displayed to the user (see Controller.syncMessage). The intention is to use that for air-gapped integration with other systems, e.g. if user submits change of status to symptoms, display message to ask user to go to a website and enter unique code to obtain test kit. It enables incremental personal data gathering, and loose integration, such that the app stays simple and private.

Your involvement in this work is very much appreciated 🤝. I'm so glad that we've finally got it to work. I suspect there are still glitches in the code, but so far it is working reliably 👍

**Comment by UsamaAbdulAziz** [comment]
Finished testing with a lot of cases and found them to be working absolutely fine. Also tested it with real world scenarios and it’s working as expected. The only downside is that if a user comes into proximity and also leaves in between screen activation notification, they won’t be able to register contact. But that it’s probability is low. 
Overall the solution is complete. Your work is absolutely fantastic here. I am also concerned if Apple is going to dislike the idea of this screen activation notification method. For that we’ll have to wait until someone tries to submit based on this solution. Anyways, bravo! 👏 👏

**Comment by c19x** [comment]
Wonderful! That is great news 👍 Thanks for doing the tests. My real world testing today has also gone perfectly.
I agree with your logic, if screen ON for both devices that have been idle for >20mins does not occur during a short encounter it won’t be registered. But the chance of that is indeed slim.
Agree about frequent notifications. Definitely need to make that a feature rather than irritant, and also need to disable notification if the screen is already on. That is my immediate next task.
Thank you so much for supporting this work 👍👍👍⭐️⭐️⭐️
Would you agree this issue is now closed? Please do stay in touch 😀

**Comment by UsamaAbdulAziz** [comment]
That would be it I guess. I can’t wait to checkout upcoming final changes. In a very few cases, contact started registering in one Phone after screen got turned on by notification, the second phone still did not register anything (although on this second screen was getting turned on every 3 minutes without any contact registration). Seems like the phone was only acting as peripheral. Will debug. I think refactoring will solve it. 

It has been very knowledgeable for me and I am glad I could be of any help. Yes, you can close the issue I guess! ✅ 👍👍👍

**Comment by c19x** [comment]
I have noticed that airplane mode ON/OFF can cause problems sometimes, that shall be investigated, and please do open a new issue if you discover other bugs and quirks. I shall also check the situation you mention regarding detection on one side only. There are still a few things I like to do to tidy up some loose ends. Shall keep you posted on progress 👍
